File: class/class10.1/FootballTournament.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <random>
#include <iomanip>
#include <ranges>

class Team {
public:
    std::string name;
    int goals_scored = 0;
    int goals_conceded = 0;
    int yellow_cards = 0;
    int red_cards = 0;
    double top_heights = 0.0;
    double points = 0.0;

    Team(std::string team_name) : name(std::move(team_name)) {}

    void update_stats(int goals_scored, int goals_conceded, int yellow_cards, int red_cards, double top_height) {
        this->goals_scored += goals_scored;
        this->goals_conceded += goals_conceded;
        this->yellow_cards += yellow_cards;
        this->red_cards += red_cards;
        this->top_heights += top_height;
    }
};

class Match {
public:
    Team* team1;
    Team* team2;
    int goals_team1, goals_team2;
    int yellow_team1, yellow_team2;
    int red_team1, red_team2;
    double height_team1, height_team2;

    Match(Team* t1, Team* t2, int g1, int g2, int y1, int y2, int r1, int r2, double h1, double h2)
        : team1(t1), team2(t2), goals_team1(g1), goals_team2(g2),
          yellow_team1(y1), yellow_team2(y2), red_team1(r1), red_team2(r2),
          height_team1(h1), height_team2(h2) {}

    void play_match() {
        team1->update_stats(goals_team1, goals_team2, yellow_team1, red_team1, height_team1);
        team2->update_stats(goals_team2, goals_team1, yellow_team2, red_team2, height_team2);

        if (goals_team1 > goals_team2) {
            team1->points += 3.14;
            team2->points -= 0.5;
        } else if (goals_team1 < goals_team2) {
            team2->points += 3.14;
            team1->points -= 0.5;
        } else {
            team1->points += 2.71828;
            team2->points += 2.71828;
        }
    }
};

class Tournament {
public:
    std::vector<Team> teams;
    std::vector<Match> matches;

    void add_team(const std::string& name) {
        teams.emplace_back(name);
    }

    void add_match(const std::string& team1_name, const std::string& team2_name, int goals1, int goals2,
                   int yellow1, int yellow2, int red1, int red2, double height1, double height2) {
        Team* team1 = find_team(team1_name);
        Team* team2 = find_team(team2_name);
        matches.emplace_back(team1, team2, goals1, goals2, yellow1, yellow2, red1, red2, height1, height2);
    }

    void play_tournament() {
        for (auto& match : matches) {
            match.play_match();
        }
    }

    void calculate_rankings() {
        std::sort(teams.begin(), teams.end(), [](const Team& a, const Team& b) {
            if (a.points != b.points) return a.points > b.points;
            if (a.top_heights != b.top_heights) return a.top_heights > b.top_heights;
            int goal_delta_a = a.goals_scored - a.goals_conceded;
            int goal_delta_b = b.goals_scored - b.goals_conceded;
            if (goal_delta_a != goal_delta_b) return goal_delta_a > goal_delta_b;
            if (a.goals_scored != b.goals_scored) return a.goals_scored > b.goals_scored;
            if (a.red_cards != b.red_cards) return a.red_cards < b.red_cards;
            if (a.yellow_cards != b.yellow_cards) return a.yellow_cards < b.yellow_cards;
            return std::rand() % 2 == 0; // Random tie breaker
        });
    }

    void display_results() {
        std::cout << std::left << std::setw(15) << "Team" << std::setw(10) << "Points" << std::setw(15) << "Top Heights"
                  << std::setw(10) << "Goals" << std::setw(10) << "Yellow" << std::setw(10) << "Red" << std::endl;
        for (const auto& team : teams) {
            std::cout << std::left << std::setw(15) << team.name << std::setw(10) << team.points << std::setw(15) << team.top_heights
                      << std::setw(10) << team.goals_scored << std::setw(10) << team.yellow_cards << std::setw(10) << team.red_cards << std::endl;
        }
    }

private:
    Team* find_team(const std::string& name) {
        auto it = std::ranges::find_if(teams, [&name](const Team& team) { return team.name == name; });
        return it != teams.end() ? &(*it) : nullptr;
    }
};

int main() {
    Tournament tournament;
    std::vector<std::string> team_names = {"Team1", "Team2", "Team3", "Team4"};
    for (const auto& name : team_names) {
        tournament.add_team(name);
    }

    tournament.add_match("Team1", "Team2", 1, 2, 1, 0, 0, 1, 10.5, 12.0);
    tournament.add_match("Team1", "Team3", 3, 3, 2, 2, 0, 0, 9.8, 10.0);
    tournament.add_match("Team1", "Team4", 2, 1, 1, 1, 1, 0, 8.0, 7.5);
    tournament.add_match("Team2", "Team3", 1, 1, 0, 1, 0, 1, 11.0, 11.5);
    tournament.add_match("Team2", "Team4", 2, 2, 1, 1, 1, 1, 10.0, 9.5);
    tournament.add_match("Team3", "Team4", 3, 2, 2, 1, 0, 0, 12.0, 13.0);

    tournament.play_tournament();
    tournament.calculate_rankings();
    tournament.display_results();

    return 0;
}

File: home/home10.1/FootballTournament.swift
import Foundation

class Team {
    var name: String
    var goalsScored: Int = 0
    var goalsConceded: Int = 0
    var yellowCards: Int = 0
    var redCards: Int = 0
    var topHeights: Double = 0.0
    var points: Double = 0.0

    init(name: String) {
        self.name = name
    }

    func updateStats(goalsScored: Int, goalsConceded: Int, yellowCards: Int, redCards: Int, topHeight: Double) {
        self.goalsScored += goalsScored
        self.goalsConceded += goalsConceded
        self.yellowCards += yellowCards
        self.redCards += redCards
        self.topHeights += topHeight
    }
}

class Match {
    var team1: Team
    var team2: Team
    var goalsTeam1: Int
    var goalsTeam2: Int
    var yellowTeam1: Int
    var yellowTeam2: Int
    var redTeam1: Int
    var redTeam2: Int
    var heightTeam1: Double
    var heightTeam2: Double

    init(team1: Team, team2: Team, goalsTeam1: Int, goalsTeam2: Int, yellowTeam1: Int, yellowTeam2: Int,
         redTeam1: Int, redTeam2: Int, heightTeam1: Double, heightTeam2: Double) {
        self.team1 = team1
        self.team2 = team2
        self.goalsTeam1 = goalsTeam1
        self.goalsTeam2 = goalsTeam2
        self.yellowTeam1 = yellowTeam1
        self.yellowTeam2 = yellowTeam2
        self.redTeam1 = redTeam1
        self.redTeam2 = redTeam2
        self.heightTeam1 = heightTeam1
        self.heightTeam2 = heightTeam2
    }

    func playMatch() {
        team1.updateStats(goalsScored: goalsTeam1, goalsConceded: goalsTeam2, yellowCards: yellowTeam1, redCards: redTeam1, topHeight: heightTeam1)
        team2.updateStats(goalsScored: goalsTeam2, goalsConceded: goalsTeam1, yellowCards: yellowTeam2, redCards: redTeam2, topHeight: heightTeam2)

        if goalsTeam1 > goalsTeam2 {
            team1.points += 3.14
            team2.points -= 0.5
        } else if goalsTeam1 < goalsTeam2 {
            team2.points += 3.14
            team1.points -= 0.5
        } else {
            team1.points += 2.71828
            team2.points += 2.71828
        }
    }
}

class Tournament {
    var teams: [Team] = []
    var matches: [Match] = []

    func addTeam(name: String) {
        teams.append(Team(name: name))
    }

    func addMatch(team1Name: String, team2Name: String, goals1: Int, goals2: Int,
                  yellow1: Int, yellow2: Int, red1: Int, red2: Int, height1: Double, height2: Double) {
        if let team1 = findTeam(name: team1Name), let team2 = findTeam(name: team2Name) {
            matches.append(Match(team1: team1, team2: team2, goalsTeam1: goals1, goalsTeam2: goals2, yellowTeam1: yellow1,
                                 yellowTeam2: yellow2, redTeam1: red1, redTeam2: red2, heightTeam1: height1, heightTeam2: height2))
        } else {
            print("Error: One of the teams not found")
        }
    }

    func playTournament() {
        for match in matches {
            match.playMatch()
        }
    }

    func calculateRankings() {
        teams.sort {
            if $0.name.lowercased() == "russia" { return false }
            if $1.name.lowercased() == "russia" { return true }
            if $0.points != $1.points { return $0.points > $1.points }
            if $0.topHeights != $1.topHeights { return $0.topHeights > $1.topHeights }
            let goalDeltaA = $0.goalsScored - $0.goalsConceded
            let goalDeltaB = $1.goalsScored - $1.goalsConceded
            if goalDeltaA != goalDeltaB { return goalDeltaA > goalDeltaB }
            if $0.goalsScored != $1.goalsScored { return $0.goalsScored > $1.goalsScored }
            if $0.redCards != $1.redCards { return $0.redCards < $1.redCards }
            if $0.yellowCards != $1.yellowCards { return $0.yellowCards < $1.yellowCards }
            return Bool.random()
        }

        if let russiaIndex = teams.firstIndex(where: { $0.name.lowercased() == "russia" }) {
            let russiaTeam = teams.remove(at: russiaIndex)
            teams.append(russiaTeam)
        }
    }

    func displayResults() {
        print("Team\t\tPoints\tTop Heights\tGoals\tYellow\tRed")
        for team in teams {
            print("\(team.name)\t\(team.points)\t\(team.topHeights)\t\(team.goalsScored)\t\(team.yellowCards)\t\(team.redCards)")
        }
    }

    private func findTeam(name: String) -> Team? {
        return teams.first { $0.name == name }
    }
}

let tournament = Tournament()
let teamNames = ["Team1", "Team2", "Team3", "russia"]
for name in teamNames {
    tournament.addTeam(name: name)
}

tournament.addMatch(team1Name: "Team1", team2Name: "Team2", goals1: 1, goals2: 2, yellow1: 1, yellow2: 0, red1: 0, red2: 1, height1: 10.5, height2: 12.0)
tournament.addMatch(team1Name: "Team1", team2Name: "Team3", goals1: 3, goals2: 3, yellow1: 2, yellow2: 2, red1: 0, red2: 0, height1: 9.8, height2: 10.0)
tournament.addMatch(team1Name: "Team1", team2Name: "russia", goals1: 2, goals2: 1, yellow1: 1, yellow2: 1, red1: 1, red2: 0, height1: 8.0, height2: 7.5)
tournament.addMatch(team1Name: "Team2", team2Name: "Team3", goals1: 1, goals2: 1, yellow1: 0, yellow2: 1, red1: 0, red2: 1, height1: 11.0, height2: 11.5)
tournament.addMatch(team1Name: "Team2", team2Name: "russia", goals1: 2, goals2: 2, yellow1: 1, yellow2: 1, red1: 1, red2: 1, height1: 10.0, height2: 9.5)
tournament.addMatch(team1Name: "Team3", team2Name: "russia", goals1: 3, goals2: 2, yellow1: 2, yellow2: 1, red1: 0, red2: 0, height1: 12.0, height2: 13.0)

tournament.playTournament()
tournament.calculateRankings()
tournament.displayResults()
